File: app/src/main/java/com/mumu/filebrowser/views/impl/ListViewImpl.java
package com.mumu.filebrowser.views.impl;

import android.animation.ObjectAnimator;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.LayerDrawable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v7.graphics.drawable.DrawableWrapper;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.util.AttributeSet;
import android.util.FloatProperty;
import android.util.Log;
import android.util.Property;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateInterpolator;
import android.widget.ImageView;
import android.widget.TextView;

import com.mumu.filebrowser.R;
import com.mumu.filebrowser.file.FileWrapper;
import com.mumu.filebrowser.file.IFile;
import com.mumu.filebrowser.views.IListView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Created by leonardo on 17-11-10.
 */

public class ListViewImpl extends RecyclerView implements IListView<IFile>, View.OnClickListener, View.OnLongClickListener {

    private static final String TAG = ListViewImpl.class.getSimpleName();

    @Nullable
    private List<IFile> mItemList;
    @NonNull
    private ListAdatpter mAdapter;
    @NonNull
    private int mLayoutStyle = LAYOUT_STYLE_LIST;
    @NonNull
    private StaggeredGridLayoutManager mLayoutManager;
    @NonNull
    private int mMode = MODE_VIEW;
    @Nullable
    private OnItemClickListener mClickListener;
    @Nullable
    private View.OnLongClickListener mLongClickListener;



    public ListViewImpl(Context context) {
        super(context);
        init();
    }

    public ListViewImpl(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public ListViewImpl(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    private void init() {
        mAdapter = new ListAdatpter();
        setAdapter(mAdapter);
        setItemAnimator(new DefaultItemAnimator());
        mLayoutManager = new StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.VERTICAL);
        setLayoutManager(mLayoutManager);
    }

    @Override
    public void setList(List<IFile> list) {
        mItemList = list;
    }

    @Nullable
    @Override
    public List<IFile> getList() {
        return mItemList;
    }

    @Override
    public void notifyDataSetChanged() {
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public void showAs(int style, boolean animate) {
        if (mLayoutStyle == style) {
            return;
        }
        mLayoutStyle = style;

        switch (style) {
            case LAYOUT_STYLE_LIST:
                mLayoutManager.setSpanCount(1);
                break;
            case LAYOUT_STYLE_GRID:
                mLayoutManager.setSpanCount(5);
                break;
        }

        if (animate) {
            //TODO add some animation here
        }
    }

    @Override
    public void showAsList(boolean anim) {
        showAs(LAYOUT_STYLE_LIST, anim);
    }

    @Override
    public void showAsGrid(boolean anim) {
        showAs(LAYOUT_STYLE_GRID, anim);
    }

    @Override
    public int getCurrentLayoutStyle() {
        return mLayoutStyle;
    }

    @Override
    public void setOnItemClickListener(@Nullable OnItemClickListener listener) {
        mClickListener = listener;
    }

    @Override
    public void switchTo(int mode, boolean anim) {
        mMode = mode;
    }

    @Override
    public void switchToViewMode(boolean anim) {
        mMode = MODE_VIEW;
    }

    @Override
    public void switchToMultiSelectMode(boolean anim) {
        mMode = MODE_MULTI_SELECT;
    }

    @Override
    public int getCurrentMode() {
        return mMode;
    }

    @Override
    public void setItemSelected(@Nullable IFile file, boolean selected) {
        checkNotNull(file);
        int target = -1;
        synchronized (mItemList) {
            target = mItemList.indexOf(file);
            View v = null;
            for(int i = getChildCount()-1;i>=0;i--){
                if((v = getChildAt(i)).getTag() == file){
                    v.setSelected(selected);
                    SimpleViewHolder holder = (SimpleViewHolder) getChildViewHolder(v);
                    holder.mDrawable.setSelected(selected, true);
                    break;
                }
            }
            if (target >= 0 && target < mAdapter.getItemCount()) {
                Log.d(TAG, "setItemSelected : target = " + target);
                ((FileWrapper) file).setSelected(selected);
            }
        }
    }

    @Override
    public void selectAll(boolean selected) {
        synchronized (mItemList) {
            for (int i = mAdapter.getItemCount() - 1; i >= 0; i--) {
                ((FileWrapper) (mItemList.get(i))).setSelected(selected);
                if (i < getChildCount()) {
                    getChildAt(i).setSelected(selected);
                }
            }
        }
    }

    @Override
    public void addItem(int positions, @NonNull IFile file) {
        synchronized (mItemList) {
            int index = positions < 0 ? 0 : positions > mItemList.size() ? mItemList.size() : positions;
            mItemList.add(index, file);
            mAdapter.notifyItemInserted(index);
            scrollToPosition(index);
        }
    }

    @Override
    public void removeItem(@NonNull IFile... items) {
        checkNotNull(items);
        synchronized (mItemList) {
            for (IFile file : items) {
                if (mItemList.contains(file)) {
                    int index = mItemList.indexOf(file);
                    mItemList.remove(file);
                    mAdapter.notifyItemRemoved(index);
                }
            }
        }
    }

    @Override
    public void onClick(View v) {
        if (mClickListener != null) {
            mClickListener.onItemClick(v, (IFile) v.getTag());
        }
    }

    @Override
    public boolean onLongClick(View v) {
        if (mClickListener != null) {
            mClickListener.onItemLongClick(v, (IFile) v.getTag());
            return true;
        }
        return false;
    }

    public class ListAdatpter extends Adapter<SimpleViewHolder> {

        @Override
        public SimpleViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext()).
                    inflate(
                            mLayoutManager.getSpanCount() == 1 ? R.layout.item_view_list : R.layout.item_view_grid,
                            parent,
                            false);
            SimpleViewHolder holder = new SimpleViewHolder(view);
            holder.mItem.setOnClickListener(ListViewImpl.this::onClick);
            holder.mItem.setOnLongClickListener(ListViewImpl.this::onLongClick);
            return holder;
        }

        @Override
        public void onBindViewHolder(SimpleViewHolder holder, int position) {
            IFile file = mItemList.get(position);
            holder.mItemName.setText(file.getName());
            Drawable icon = DrawableCompat.wrap(file.getIcon(getResources()));
            Drawable selectedDrawable = getResources().getDrawable(R.drawable.ic_item_selected, null);
            SelectDrawable drawable = new SelectDrawable(
                    new Drawable[]{
                            icon.mutate(),
                            selectedDrawable.mutate()
                    });
            holder.setIconDrawable(drawable);
            boolean selected = ((FileWrapper) file).isSelected();
            drawable.setSelected(selected, false);
            holder.mItem.setSelected(selected);
            holder.mItem.setTag(file);
        }

        @Override
        public int getItemCount() {
            return mItemList == null ? 0 : mItemList.size();
        }

        @Override
        public int getItemViewType(int position) {
            return mLayoutStyle;
        }
    }

    class SimpleViewHolder extends ViewHolder {
        @BindView(R.id.item)
        View mItem;
        @BindView(R.id.item_name)
        TextView mItemName;
        @BindView(R.id.item_icon)
        ImageView mItemIcon;
        SelectDrawable mDrawable;

        public SimpleViewHolder(@NonNull View itemView) {
            super(itemView);
            checkNotNull(itemView);
            ButterKnife.bind(this, itemView);
        }

        public void setIconDrawable(@NonNull SelectDrawable drawable) {
            checkNotNull(drawable);
            mDrawable = drawable;
            mItemIcon.setImageDrawable(mDrawable);
        }
    }

    class SelectDrawable extends LayerDrawable {

        float animValue;

        private Property<SelectDrawable, Float> ALPHA_PROPERTY = new FloatProperty<SelectDrawable>(TAG) {
            @Override
            public Float get(SelectDrawable object) {
                return object.animValue;
            }

            @Override
            public void setValue(SelectDrawable object, float value) {
                object.setAnimValue(value);
            }
        };

        private SelectDrawable(@NonNull Drawable[] layers) {
            super(layers);
        }

        private void setAnimValue(float value) {
            if (value < 0) value = 0;
            if (value > 1f) value = 1f;
            animValue = value;
            getDrawable(0).setAlpha((int) ((1f - value) * 255f));
            getDrawable(1).setAlpha((int) (value * 255f));
            invalidateSelf();
        }

        public void setSelected(boolean selected, boolean needAnim) {
            if (needAnim) {
                float start = selected ? 0 : 1f;
                ObjectAnimator anim = ObjectAnimator
                        .ofFloat(this, ALPHA_PROPERTY, start, 1f - start)
                        .setDuration(getResources().getInteger(R.integer.item_icon_animation_duration));
                anim.setInterpolator(new AccelerateInterpolator());
                anim.start();
            } else {
                setAnimValue(selected ? 1 : 0);
            }
        }
    }
}

File: app/src/main/java/com/mumu/filebrowser/views/impl/ListViewImpl.kt
package com.mumu.filebrowser.views.impl

import android.animation.ObjectAnimator
import android.content.Context
import android.graphics.drawable.Drawable
import android.graphics.drawable.LayerDrawable
import android.support.v4.graphics.drawable.DrawableCompat
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.StaggeredGridLayoutManager
import android.util.AttributeSet
import android.util.FloatProperty
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateInterpolator
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import com.mumu.filebrowser.R
import com.mumu.filebrowser.views.IListView
import com.mumu.filebrowser.presenter.IListPresenter
import com.mumu.filebrowser.presenter.IPresenter
import com.mumu.filebrowser.presenter.impl.ListPresenterImpl
import com.mumu.filebrowser.utils.PathUtils

class ListViewImpl : FrameLayout, IListView<String>, View.OnClickListener, View.OnLongClickListener {
    companion object {
        private val sListPresenter: IListPresenter<String> = ListPresenterImpl()
    }

    private var mEmptyView: View? = null
    private var mRecyclerView: RecyclerView? = null

    constructor(context: Context) : super(context)
    constructor(context: Context, attributes: AttributeSet?) : super(context, attributes)
    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle)

    private fun init(view: RecyclerView) {
        mRecyclerView = view
        mRecyclerView?.adapter = ListAdapter()
        mRecyclerView?.itemAnimator = DefaultItemAnimator()
        mRecyclerView?.layoutManager = StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.VERTICAL)
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        init(findViewById(R.id.main_list))
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        (sListPresenter as IPresenter).bindView(this)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        (sListPresenter as IPresenter).bindView(null)
    }

    override fun onClick(v: View?) {
        sListPresenter?.onItemClick(v?.tag as String)
    }

    override fun onLongClick(v: View?): Boolean {
        sListPresenter?.onItemLongClick(v?.tag as String)
        return true
    }

    override fun showAsList(anim: Boolean) {
        (mRecyclerView?.layoutManager as StaggeredGridLayoutManager).spanCount = 1
    }

    override fun showAsGrid(anim: Boolean) {
        (mRecyclerView?.layoutManager as StaggeredGridLayoutManager)?.spanCount = 6
    }

    override fun notifyDataSetChanged() {
        //keep in main-thread
        handler.post { mRecyclerView?.adapter?.notifyDataSetChanged() }
    }

    override fun notifyItemInserted(index: Int) {
        //keep in main-thread
        handler.post { mRecyclerView?.adapter?.notifyItemInserted(index) }
    }

    override fun notifyItemRemoved(index: Int) {
        //keep in main-thread
        handler.post { mRecyclerView?.adapter?.notifyItemRemoved(index) }
    }

    override fun setEmptyView(v: View) {
        //keep in main-thread
        handler.post {
            if (mEmptyView != null) {
                removeViewAt(0)
            }
            mEmptyView = v
            addView(mEmptyView, 0)
        }
    }

    override fun setEmptyView(layout: Int) {
        setEmptyView(View.inflate(context, layout, null))
    }

    override fun select(items: String?, selected: Boolean) {
        if (items == null) {
            for (i in 0..(mRecyclerView!!.childCount - 1)) {
                val view = mRecyclerView!!.getChildAt(i)
                if (!view.isSelected) {
                    continue
                }
                view.isSelected = false
                val holder = mRecyclerView?.getChildViewHolder(view) as SimpleViewHolder
                holder.mDrawable?.setSelected(false, true)
            }
            return
        }
        for (i in 0..(mRecyclerView!!.childCount - 1)) {
            val view = mRecyclerView!!.getChildAt(i)
            if (items == view.tag) {
                view.isSelected = selected
                val holder = mRecyclerView?.getChildViewHolder(view) as SimpleViewHolder
                holder.mDrawable?.setSelected(selected, true)
                break
            }
        }
    }

    override fun focus(item: String, focus: Boolean) {
        for (i in 0..(mRecyclerView!!.childCount - 1)) {
            val view = mRecyclerView!!.getChildAt(i)
            if (focus) view.requestFocus() else view.clearFocus()
            break
        }

    }

    /*Adapter*/
    private inner class ListAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
            val view = LayoutInflater.from(parent.context).inflate(
                    if ((mRecyclerView?.layoutManager as StaggeredGridLayoutManager)?.spanCount == 1) R.layout.item_view_list else R.layout.item_view_grid,
                    parent,
                    false)
            val holder = SimpleViewHolder(view)
            holder.mItem!!.setOnClickListener { this@ListViewImpl.onClick(it) }
            holder.mItem!!.setOnLongClickListener { this@ListViewImpl.onLongClick(it) }
            return holder
        }

        override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
            if (holder is SimpleViewHolder) {
                val name = sListPresenter!!.list.get(position)
                Log.i("onBindViewHolder", "onBindViewHolder -> bind $name")
                val file = PathUtils.get(name)
                holder.mItemName!!.text = file.getName()
                val icon = DrawableCompat.wrap(file.getIcon(resources))
                val selectedDrawable = resources.getDrawable(R.drawable.ic_item_selected, null)
                val drawable = SelectDrawable(
                        arrayOf(icon.mutate(), selectedDrawable.mutate()))
                holder.setDrawable(drawable)
                val selected = sListPresenter.isItemSelected(name)
                drawable.setSelected(selected, false)
                holder.mItem!!.isSelected = selected
                holder.mItem!!.tag = name
            }
        }

        override fun getItemCount(): Int {
            val size = sListPresenter?.list.size
            mEmptyView?.visibility = if (size == 0) View.VISIBLE else View.GONE
            return size
        }

        override fun getItemViewType(position: Int): Int {
            return sListPresenter?.currentLayoutStyle
        }
    }

    internal inner class SimpleViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val mItem: View = itemView.findViewById(R.id.item)
        val mItemName: TextView = itemView.findViewById(R.id.item_name)
        val mItemIcon: ImageView = itemView.findViewById(R.id.item_icon)
        var mDrawable: SelectDrawable? = null
        fun setDrawable(drawable: SelectDrawable) {
            mDrawable = drawable
            mItemIcon.setImageDrawable(mDrawable)
        }
    }

    internal inner class SelectDrawable constructor(layers: Array<Drawable>) : LayerDrawable(layers) {
        var animValue: Float = 0F
            set(value) {
                var v = value
                if (value < 0) v = 0f
                if (value > 1f) v = 1f
                getDrawable(0).alpha = ((1f - v) * 255f).toInt()
                getDrawable(1).alpha = (v * 255f).toInt()
                invalidateSelf()
            }

        private val ALPHA_PROPERTY = object : FloatProperty<SelectDrawable>("ALPHA_PROPERTY") {
            override fun get(obj: SelectDrawable) = obj.animValue

            override fun setValue(obj: SelectDrawable, value: Float) {
                obj.animValue = value
            }
        }

        fun setSelected(selected: Boolean, needAnim: Boolean) = if (needAnim) {
            val start: Float = if (selected) 0f else 1f
            val anim = ObjectAnimator
                    .ofFloat<SelectDrawable>(this, ALPHA_PROPERTY, start, 1f - start)
                    .setDuration(resources.getInteger(R.integer.item_icon_animation_duration).toLong())
            anim.interpolator = AccelerateInterpolator()
            anim.start()
        } else {
            animValue = if (selected) 1f else 0f
        }
    }
}